File: wan2good/프로그래머스/31주차_모두0으로만들기.java
import java.util.*;

class Solution {
    public static long solution(int[] a, int[][] edges) {
        long answer = 0;
        long sum = 0;
        long[] cost = new long[a.length];
        
        for(int i=0; i<a.length; i++) {
        	sum+=a[i];
        	cost[i] = a[i];
        }
        
        if(sum!=0)
        	return -1;
        
        int[] indegree = new int[a.length];
        boolean[] visited = new boolean[a.length];
        ArrayList<Integer>[] graph = new ArrayList[a.length];
        for(int i=0; i<a.length; i++) {
        	graph[i] = new ArrayList<>();
        }
        
        for(int[] edge : edges) {
        	graph[edge[0]].add(edge[1]);
        	graph[edge[1]].add(edge[0]);
        	indegree[edge[0]]++;
        	indegree[edge[1]]++;
        }
        
        Queue<Integer> q = new LinkedList<>();
        for(int i=0; i<a.length; i++) {
        	if(indegree[i]==1)
        		q.offer(i);
        }
        
        while(!q.isEmpty()) {
        	int cur = q.poll();
        	visited[cur] = true;
        	
        	for(int parent : graph[cur]) {
        		if(!visited[parent]) {
        			answer += Math.abs(cost[cur]);
        			cost[parent]+=cost[cur];
        			cost[cur]=0;
        			if(--indegree[parent]==1)
        				q.offer(parent);
        		}
        	}
        }
        
        return answer;
    }
}

public class Main {
	public static void main(String[] args){
		
		Solution solution = new Solution();
		int[] a = {-5,0,2,1,2};
		int[][] edges = {{0,1},{3,4},{2,3},{0,3}};
		System.out.println(solution.solution(a, edges));
		
	}
}

File: JJangSungWon/[3]Programmers/31주차_모두0으로만들기.py
import sys

sys.setrecursionlimit(10 ** 8)


def dfs(idx, a):
    global visited, graph, answer
    now = a[idx]
    visited[idx] = True

    for index in graph[idx]:
        if visited[index] is False:
            now += dfs(index, a)
    answer += abs(now)
    return now


def solution(a, edges):
    global visited, graph, answer
    answer = 0

    if sum(a) != 0:
        return -1

    visited = [False] * len(a)
    graph = [[] for _ in range(len(a))]

    for x, y in edges:
        graph[x].append(y)
        graph[y].append(x)

    dfs(0, a)

    return answer